docs/design/proposals/collections.md

```diff
@@ -0,0 +1,633 @@
+                # collection resource can be referenced by its ID
+                paths:
+                  | # kro will do it's magic making sure to replace this with the proper type.
+                  ${microservices.map(svc, {
+                    'path': '/' + svc.metadata.labels.app,
+                    'pathType': 'Prefix',
+                    'backend': {
+                      'service': {
+                        'name': svc.metadata.name,
+                        'port': {'number': 80}
+                      }
+                    }
+                  })}
+```
+
+or you might want to reference a collection to create another collection
+(collection piping):
+
+```yaml
+kind: ResourceGraphDefinition
+# ...
+spec:
+  resources:
+    # Create a collection of pods for each database
+    - id: databases
+      forEach: ${spec.databases}
+      template:
+        kind: Pod
+        # ...
+
+    - id: databaseBackups
+      # Create a backup job for each database
+      forEach: ${databases}
+      template:
+        kind: CronJob
+        metadata:
+          name: ${each.item.metadata.name + '-backup'}
+        spec:
+          schedule: "0 2 * * *"
+          jobTemplate:
+            # ...
+```
+
+### External references with collections
+
+Today, KRO supports `externalRef` to discover resources outside the RGD.
+external references does point to one and unique resource, discovered by name
+and namespace.
+
+The `externalRef` field can be used in conjunction with collections to
+dynamically discover resources and create new resources based on the discovered
+set. For example, you might want to create a Pod for each Namespace that matches
+a certain label:
+
+#### Discover resources based on labels
+
+```yaml
+- id: targetNamespaces
+  externalRef:
+    apiVersion: v1
+    kind: Namespace
+    selector:
+      matchLabels:
+        team: frontend
+```
+
+#### Discover resources using forEach semantics
+
+```yaml
+- id: namespacePods
+  forEach: ${targetNamespaces} # Iterate over discovered namespaces
+  template:
+    kind: Pod
+    metadata:
+      name: ${'pod-in-' + each.item.metadata.name}
+      namespace: ${each.item.metadata.name}
+```
+
+## Scope
+
+This proposal focuses on enabling dynamic resource creation through the
+`forEach` field and essential supporting features.
+
+### In Scope
+
+- The `forEach` field for iterating over arrays, maps, and resource collections
+- Iteration context variables (`each.item`, `each.index`, `each.key`,
+  `each.value`, `each.length`)
+- Resource references in CEL to enable collection chaining
+- Status field computations using collections
+- Integration with existing KRO fields (`readyWhen`, `includeWhen`)
+- Updates to resources when collection items change
+- Enhanced externalRef to watch collections of resources (via slectors)
+- Combining `forEach` with externalRef for dynamic external resource discovery
+
+### Out of Scope
+
+- Nested forEach loops or recursive expansion
```

Comment:
As mentioned, I think if we declare this out of scope we should at least:
- Have a plan for how we could (in a nonbreaking way) expand to support it in the future.
- Show how you can do cartesian product w/ maps/lists in CEL natively, meaning users can easily work around this lack.